### DISCUSSION TITLE
Autocompletion issues when using VSCode Workspaces

### DISCUSSION THREAD

**Discussion by Alain-Godo** [BradenM/micropy-cli#471] [Q&A] [answered]
First of all, I have to congratulate you for this beautiful project !!!. I followed every step in the guide, and all works fine...but when the project is created and I try to write my code, Pylance still does not detect the modules. Why?

![Error](https://github.com/BradenM/micropy-cli/assets/125923673/2c94bdb1-f8b1-4a20-9818-c68da866f827)

**Comment by BradenM** [comment]
Hi @Alain-Godo, thanks for the report.

Will some information before I can provide any help, though.

Would you mind sharing your `micropy --version`, `micropy.json` file and `~/.micropy/micropy.log` files, please? Thanks.

**Reply by Alain-Godo** [reply]
Sure !!! But I can't find the ~/.micropy/micropy.log file.

![image](https://github.com/BradenM/micropy-cli/assets/125923673/9115a642-f4ab-4ce2-bb29-7131c20a4824)

**Reply by BradenM** [reply]
Thanks.

Although I have no reason to believe the `esp32-micropython-1.15.0` stubs shouldn't work -- I would recommend trying to use a package from the newer micropython-stubs repository:
`micropy stubs add micropython-esp32-stubs` to add latest version (see `micropy stubs search --show-outdated esp32` for older versions you can specify in micropy.json). These stubs are much more up to date and provide richer autocompletion and docstrings.

Since I still haven't gotten around to implementing a simple `stubs remove` command :grimacing: -- you can remove the stubs you are currently using by removing the entry from `micropy.json` and recreating the `.micropy` folder (`rm -rf .micropy && micropy`)

Please let me know if this works for you, as it may indicate a bug in how micropy handles the older stub packages if so. Thanks.

**Reply by Alain-Godo** [reply]
I tried all from the being...and I selected the new two stubs (because I don't know what is the right one ), see the micropy.json...I don't know what is wrong : (... Sorry if my English is not the best....Thanks

![image](https://github.com/BradenM/micropy-cli/assets/125923673/cbdf46e4-bd40-4ccc-9cc8-ac1805bd3769)

![image](https://github.com/BradenM/micropy-cli/assets/125923673/92320f38-952b-421c-8122-cabd93e67803)

![image](https://github.com/BradenM/micropy-cli/assets/125923673/9a1bce67-4596-469f-ba45-96f21b235813)

**Reply by BradenM** [reply]
@Alain-Godo Selecting both is correct. :thinking:

Would you mind sharing your `.vscode/settings.json` file?
Also, which version of python are you using here?

**Reply by Alain-Godo** [reply]
Sure!! 
Version of Python: 3.11.2 - 64 bits

.vscode/settings.json:
-------------------------------------------------------------------------------------------------------------------------------------------
{
    //Loaded Stubs:  micropython-esp32-stubs  micropython-stdlib-stubs 
    
    "python.languageServer": "Pylance",
    "python.analysis.autoSearchPath": true,
    "python.autoComplete.extraPaths": [".micropy\\micropython_esp32_stubs-1.19.1.post12", ".micropy\\micropython-stdlib-stubs-0.9.0", ".micropy\\TinySorbaTest"],
    
    

    "python.linting.enabled": true,
    "python.linting.pylintEnabled": true
}
-------------------------------------------------------------------------------------------------------------------------------------------

**Reply by BradenM** [reply]
@Alain-Godo Thanks for the info. I've done some digging, and it seems the configurations micropy generates are now outdated as of more recent vscode/pylance versions. I am working on a fix now. See: #474
Thanks for bringing this to my attention.

**Comment by BradenM** [comment]
@Alain-Godo I believe this should be resolved now as of the v4.2.1 version that I've just released.
Thanks again for bringing this to my attention!
Please let me know if this isn't the case.

**Reply by Alain-Godo** [reply]
I updated the package and created a new project... Nothing : (

{
    // Loaded Stubs:  micropython-stdlib-stubs  micropython-esp32-stubs 
    
    "python.languageServer": "Pylance",
    "python.analysis.autoSearchPath": true,
    "python.autoComplete.extraPaths": [".micropy\\micropython-stdlib-stubs-0.9.0\\stdlib", ".micropy\\micropython_esp32_stubs-1.19.1.post12", ".micropy\\tinysorba_esp32"],
    "python.analysis.diagnosticSeverityOverrides": { "reportMissingModuleSource": "none" },
    "python.analysis.typeCheckingMode": "basic",
    "python.autoComplete.typeshedPaths":  [".micropy\\micropython-stdlib-stubs-0.9.0\\stdlib", ".micropy\\micropython_esp32_stubs-1.19.1.post12", ".micropy\\tinysorba_esp32", "typings"],
    "python.analysis.typeshedPaths":  [".micropy\\micropython-stdlib-stubs-0.9.0\\stdlib", ".micropy\\micropython_esp32_stubs-1.19.1.post12", ".micropy\\tinysorba_esp32", "typings"],
    
    

    "python.linting.enabled": true,
    "python.linting.pylintEnabled": true
}

**Reply by BradenM** [reply]
@Alain-Godo What behavior are you now seeing with the latest version?
I started a windows 11 vm, installed vscode, python, micropy, and pylint and am unable to replicate this.

![image](https://github.com/BradenM/micropy-cli/assets/5913808/e9c1d44a-dbb9-44a1-abd8-9b4ebb22fd2f)

**Reply by Alain-Godo** [reply]
The only difference I see is that I have Windows 10. Well, and my Python version is 3.11.2 instead of 3.11.1 u are using, just that.

**Reply by Alain-Godo** [reply]
Wait I see something.... Look

![image](https://github.com/BradenM/micropy-cli/assets/125923673/62966db8-b82a-426f-8704-12f2a91b361a)

![image](https://github.com/BradenM/micropy-cli/assets/125923673/2a798890-8d3e-42aa-938c-3cdc6585d4f7)

**Reply by BradenM** [reply]
@Alain-Godo The first image can be ignored. It's a reference to an older setting and just there for compatibility with older versions (probably can be removed in near future.).

The second image is interesting and definitely seems problematic.
Can you try opening the project directly as it suggests? I.e, by running `code .` from the directory of your project?

**Reply by Alain-Godo** [reply]
It works!!!!!!!!! : ) But what happened?? 

![image](https://github.com/BradenM/micropy-cli/assets/125923673/f1ec970e-b7c9-4244-a11d-dc45652ef5de)

**Reply by BradenM** [reply, marked as answer]
Interesting. So the issue looks to be something to do with vscode workspaces (which is for working on multiple projects in the same window). I will have to investigate this, but from your images it looks like pylance just isn't loaded in workspaces and you need to open the project directly (which is what `code .` did for you) instead. I will have to read into why this is.

**Reply by Alain-Godo** [reply]
I hope it can be solved, now it's working great!!! It is really a super necessary project that you do, I congratulate you. Thanks for all !!!!

**Reply by Alain-Godo** [reply]
One last question !!! How can I use the pymakr.conf to upload the code to my device? I read this post, but it's not clear enough: https://lemariva.com/blog/2019/08/micropython-vsc-ide-intellisense

**Reply by BradenM** [reply]
@Alain-Godo Glad to hear you've got a working setup now.

As for pymakr, it's been quite a while since I have used it, but I do know that micropy currently generates a config for the older v1 version (see: #312). 

As mentioned in #312, pymakr seems to be more or less discontinued and in the near future I am working on implementing this functionality in micropy itself. 

But for now, I would look through that issue and hopefully, you will find what you need. Thanks.
